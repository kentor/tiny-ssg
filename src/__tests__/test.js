const path = require('path');
const stripAnsi = require('strip-ansi');

let mockOutputFile;
let mockRemove;

jest.mock('fs-extra', () => ({
  outputFile: mockOutputFile,
  remove: mockRemove,
}));

function withoutBuildTime(logs) {
  return logs
    .map(args => args.map(stripAnsi))
    .filter(args => !args[0].startsWith('built'));
}

describe('ssg', () => {
  let deletes;
  let logs;
  let SSG;
  let ssg;
  let writes;

  beforeEach(() => {
    deletes = [];
    logs = [];
    writes = [];

    mockOutputFile = jest.fn(async (outFile, data) => {
      writes.push([outFile, data]);
    });

    mockRemove = jest.fn(async (outFile) => {
      deletes.push(outFile);
    });

    console.log = (...args) => {
      logs.push(args);
    };

    SSG = require('../index');
    ssg = new SSG({ dest: '/public/' });
  });

  describe('#build', () => {
    it('throws when called without first calling manifest', async () => {
      try {
        await ssg.build();
      } catch (err) {
        expect(err.message).toMatch(/build.*?manifest/);
      }
    });

    it('writes according to dest and manifest', async () => {
      ssg.manifest(() => [
        { url: '/', view: () => 'index page' },
        { url: '/about/', view: async () => 'about page' },
        { url: '/feed.xml', view: () => '<?xml>feed</xml>' },
      ]);

      await ssg.build();

      expect(writes).toMatchSnapshot();
      expect(withoutBuildTime(logs)).toMatchSnapshot();
    });

    it('does not write the same content twice', async () => {
      ssg.manifest(() => [{ url: '/', view: () => '' }]);
      await ssg.build();
      expect(writes).toMatchSnapshot();
      await ssg.build();
      expect(writes).toMatchSnapshot();
    });

    it('updates a page when view function returns a new value', async () => {
      ssg.manifest(() => [{ url: '/', view: () => '1' }]);
      await ssg.build();
      expect(writes).toMatchSnapshot();
      ssg.manifest(() => [{ url: '/', view: () => '2' }]);
      await ssg.build();
      expect(writes).toMatchSnapshot();
    });

    it('deletes a page when it is no longer in manifest', async () => {
      ssg.manifest(() => [
        { url: '/', view: () => '' },
        { url: '/about/', view: () => '' },
      ]);
      await ssg.build();
      expect(writes).toMatchSnapshot();
      ssg.manifest(() => [
        { url: '/', view: () => '' },
      ]);
      await ssg.build();
      expect(deletes).toMatchSnapshot();
      expect(withoutBuildTime(logs)).toMatchSnapshot();
    });

    it('passes meta and a cloned page object to the view', async () => {
      const meta = {};
      ssg.manifest(() => [{
        url: '/',
        view: (meta, page) => {
          expect(meta).toBe(meta);
          expect(page.url).toBe('/');
          return '';
        },
        meta,
      }]);
      await ssg.build();
    });

    it('is debounced', async () => {
      ssg.manifest(() => []);

      const first = ssg.build();
      const second = ssg.build();
      expect(second).toBe(first);

      await new Promise(resolve => setTimeout(resolve, 1));

      const third = ssg.build();
      expect(third).not.toBe(first);
    });

    it('warns when a page does not follow the expected shape', async () => {
      ssg.manifest(() => [
        { url: '/missing-view/' },
        { view: () => 'missing url' },
        'something random',
      ]);
      await ssg.build();
      expect(withoutBuildTime(logs)).toMatchSnapshot();
    });

    it('warns when a url is encountered twice', async () => {
      ssg.manifest(() => [
        { url: '/test/', view: () => '1' },
        { url: '/test/', view: () => '2' },
      ]);
      await ssg.build();
      expect(withoutBuildTime(logs)).toMatchSnapshot();
      expect(writes).toMatchSnapshot();
    });
  });

  describe('#dryRun', () => {
    it('logs out the urls from manfiest', () => {
      ssg.manifest(() => [
        { url: '/', view: () => '' },
        { url: '/about/', view: () => '' },
        { url: '/feed.xml', view: () => '' },
      ]);
    });
  });
});


describe('instance', () => {
  it('is an instance of ssg', () => {
    const instance = require('../instance');
    const SSG = require('../index');
    expect(instance).toBeInstanceOf(SSG);
    expect(instance.dest).toBe(path.resolve('public'));
  });
});
