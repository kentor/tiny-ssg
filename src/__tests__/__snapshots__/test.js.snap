// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ssg #build deletes a page when it is no longer in manifest 1`] = `
Array [
  Array [
    "/public/index.html",
    "",
  ],
  Array [
    "/public/about/index.html",
    "",
  ],
]
`;

exports[`ssg #build deletes a page when it is no longer in manifest 2`] = `
Array [
  "/public/about/index.html",
]
`;

exports[`ssg #build deletes a page when it is no longer in manifest 3`] = `
Array [
  Array [
    "wrote",
    "/",
    0,
  ],
  Array [
    "wrote",
    "/about/",
    0,
  ],
  Array [
    "deleted",
    "/about/",
  ],
]
`;

exports[`ssg #build does not write the same content twice 1`] = `
Array [
  Array [
    "/public/index.html",
    "",
  ],
]
`;

exports[`ssg #build does not write the same content twice 2`] = `
Array [
  Array [
    "/public/index.html",
    "",
  ],
]
`;

exports[`ssg #build updates a page when view function returns a new value 1`] = `
Array [
  Array [
    "/public/index.html",
    "1",
  ],
]
`;

exports[`ssg #build updates a page when view function returns a new value 2`] = `
Array [
  Array [
    "/public/index.html",
    "1",
  ],
  Array [
    "/public/index.html",
    "2",
  ],
]
`;

exports[`ssg #build warns when a page does not follow the expected shape 1`] = `
Array [
  Array [
    "warn",
    Object {
      "url": "/missing-view/",
    },
    "should have shape",
    "{ url: string, view: () => Promise<string> | string }",
  ],
  Array [
    "warn",
    Object {
      "view": [Function],
    },
    "should have shape",
    "{ url: string, view: () => Promise<string> | string }",
  ],
  Array [
    "warn",
    "something random",
    "should have shape",
    "{ url: string, view: () => Promise<string> | string }",
  ],
]
`;

exports[`ssg #build warns when a url is encountered twice 1`] = `
Array [
  Array [
    "warn",
    "duplicate entry found for /test/. skipping...",
  ],
  Array [
    "wrote",
    "/test/",
    1,
  ],
]
`;

exports[`ssg #build warns when a url is encountered twice 2`] = `
Array [
  Array [
    "/public/test/index.html",
    "1",
  ],
]
`;

exports[`ssg #build writes according to dest and manifest 1`] = `
Array [
  Array [
    "/public/index.html",
    "index page",
  ],
  Array [
    "/public/about/index.html",
    "about page",
  ],
  Array [
    "/public/feed.xml",
    "<?xml>feed</xml>",
  ],
]
`;

exports[`ssg #build writes according to dest and manifest 2`] = `
Array [
  Array [
    "wrote",
    "/",
    10,
  ],
  Array [
    "wrote",
    "/about/",
    10,
  ],
  Array [
    "wrote",
    "/feed.xml",
    16,
  ],
]
`;
